syncer/atxsync/syncer.go

```diff
@@ -0,0 +1,329 @@
+	EpochInfoPeers int `mapstructure:"epoch-info-peers"`
+
+	// RequestsLimit is the maximum number of requests for single activation.
+	//
+	// The purpose of it is to prevent peers from advertising invalid atx and disappearing.
+	// Which will make node ask other peers for invalid atx.
+	// It will be reset to 0 once atx advertised again.
+	RequestsLimit int `mapstructure:"requests-limit"`
+
+	// AtxsBatch is the maximum number of atxs to sync in a single request.
+	AtxsBatch int `mapstructure:"atxs-batch"`
+}
+
+func WithConfig(cfg Config) Opt {
+	return func(s *Syncer) {
+		s.cfg = cfg
+	}
+}
+
+func New(fetcher fetcher, clock clock, db sql.Executor, localdb *localsql.Database, opts ...Opt) *Syncer {
+	s := &Syncer{
+		logger:  zap.NewNop(),
+		cfg:     DefaultConfig(),
+		fetcher: fetcher,
+		clock:   clock,
+		db:      db,
+		localdb: localdb,
+	}
+	for _, opt := range opts {
+		opt(s)
+	}
+	return s
+}
+
+type Syncer struct {
+	logger  *zap.Logger
+	cfg     Config
+	fetcher fetcher
+	clock   clock
+	db      sql.Executor
+	localdb *localsql.Database
+}
+
+func (s *Syncer) closeToTheEpoch(publish types.EpochID, timestamp time.Time) bool {
+	target := publish + 1
+	epochStart := s.clock.LayerToTime(target.FirstLayer())
+	return timestamp.After(epochStart) || epochStart.Sub(timestamp) < 2*s.cfg.EpochInfoInterval
+}
+
+func (s *Syncer) Download(ctx context.Context, publish types.EpochID) error {
+	s.logger.Info("starting atx sync", log.ZContext(ctx), publish.Field().Zap())
+
+	state, err := atxsync.GetSyncState(s.localdb, publish)
+	if err != nil {
+		return fmt.Errorf("failed to get state for epoch %v: %w", publish, err)
+	}
+	lastSuccess, err := atxsync.GetRequestTime(s.localdb, publish)
+	if err != nil && !errors.Is(err, sql.ErrNotFound) {
+		return fmt.Errorf("failed to get last request time for epoch %v: %w", publish, err)
+	}
+	// in case of immediate we will request epoch info without waiting EpochInfoInterval
+	immediate := len(state) == 0 || (errors.Is(err, sql.ErrNotFound) || !s.closeToTheEpoch(publish, lastSuccess))
+
+	ctx, cancel := context.WithCancel(ctx)
+	eg, ctx := errgroup.WithContext(ctx)
+	updates := make(chan epochUpdate, s.cfg.EpochInfoPeers)
+	if len(state) == 0 {
+		state = map[types.ATXID]int{}
+	} else {
+		updates <- epochUpdate{time: lastSuccess, update: state}
+	}
+	// termination requires two conditions:
+	// - epoch info has to be successfully downloaded close to or after the epoch start
+	// - all atxs from that epoch have to be downloaded or they are unavailable.
+	//   atx is unavailable if it was requested more than RequestsLimit times, and no peer provided it.
+	eg.Go(func() error {
+		return s.downloadEpochInfo(ctx, publish, immediate, updates)
+	})
+	eg.Go(func() error {
+		err := s.downloadAtxs(ctx, publish, state, updates)
+		cancel()
+		return err
+	})
+	return eg.Wait()
+}
+
+func (s *Syncer) downloadEpochInfo(
+	ctx context.Context,
+	publish types.EpochID,
+	immediate bool,
+	updates chan epochUpdate,
```

Comment:
NIT to state the intent (how it will be used - as a sender)
```suggestion
	updates chan<- epochUpdate,
```